design/design-doc-observe-only-resources.md

```diff
@@ -0,0 +1,703 @@
+	Update(ctx context.Context, mg resource.Managed) (ExternalUpdate, error)
+	Delete(ctx context.Context, mg resource.Managed) error
+}
+```
+
+We will leverage the fact that we have an already implemented Observe method for
+all managed resources by calling only it when the Management Policy is set to
+`ObserveOnly`. This will require minor modifications in the Managed Reconciler
+code (in the Crossplane Runtime) that will return early in the reconcile loop
+and prevent invocation of the other methods, namely, Create, Update and Delete.
+These modifications will implement the following logic at a high level:
+
+Right after the `Observe` method invocation, if `ObserveOnly`:
+
+- Return error if the resource does not exist.
+- Publish connection details.
+- Ignore late-initialization result and never call `client.Update` method to
+  update the resource spec.
+- Report success and return early.
+
+We will also need the following changes per resource:
+
+- Update the API schema to have all the fields under `spec.forProvider` under
+  `status.atProvider` as well.
+- Update the `Observe` method implementation to populate the `status.atProvider`
+  with the full state of the resource.
+
+#### Feature Gating
+
+Similar to all other new features being added to Crossplane, we will ship this
+new policy as an alpha feature that will be off by default and will be
+controlled by `--enable-alpha-management-policies` flag in Providers.
+
+This will not prevent the field from appearing in the schema of the managed
+resources. However, we will ignore the `spec.managementPolicy` when the feature
+is not enabled.
+
+#### Deprecation of `deletionPolicy`
+
+With the new `managementPolicy` covering the existing `deletionPolicy`, we will
+deprecate the latter in favor of the former.
+
+Until we drop the `deletionPolicy` from the schema altogether, we need to be
+careful with the conflicting combinations shown in the below table which only
+exists with deletion:
+
+| Deletion Policy | Management Policy | Should Observe? | Should Create? | Should Update? | Should Delete? |
+| --- | --- | --- | --- | --- | --- |
+| Delete | Full | Yes | Yes | Yes | Yes |
+| Orphan | OrphanOnDelete | Yes | Yes | Yes | No |
+| Delete | ObserveOnly | Yes | No | No | Conflict (No) |
+| Orphan | Full | Yes | Yes | Yes | Conflict (No) |
+| Delete | OrphanOnDelete | Yes | Yes | Yes | Conflict (No) |
+| Orphan | ObserveOnly | Yes | No | No | No |
+
+For conflicting cases, we will decide based on the non-default configuration
+which means "not deleting the external resource" for all 3 conflicting cases.
+This way, we will also err on the side of caution by leaving the actual resource
+untouched, avoiding any accidental deletion or modification.
+
+> Another solution could be simply throwing an error and preventing
+> reconciliation during conflict. This would be more explicit but would require
+> some manual actions and degraded UX for the usage of the feature, for example:
```

Comment:
The only difference from the current proposal would be **deleting the external resource** when `DeletionPolicy: Orphan && ManagementPolicy: Full` instead of orphaning it. This could happen when someone had intentionally set `DeletionPolicy: Orphan` on existing resources and later enabled the feature flag without touching the ManagementPolicy field on existing resources (ending with its default value). I wouldn't be comfortable deleting external resources in this case and would rather orphan it obeying the non-default configuration reflecting the user's intent.

FWIW, while the table might look complicated, this is how it looks like when implemented: https://github.com/crossplane/crossplane-runtime/pull/384/files#diff-97300a2543f95f5a2ada3560bf47dd7334e237e27976574d15d1cddef2e66c01R1105